### PR TITLE
CI: use nextest for running rust tests

### PR DESCRIPTION
**Pull request by coszio**
## Motivation

Our current CI test scheme uses the integrated `cargo test` tool for running tests, but sometimes tests begin to flake out, which makes us retry the workflow run to make sure it wasn't our fault. Arguably this makes us pay more attention to flaky tests, but it becomes a distraction when doing unrelated work.

Recently I found out about [`nextest`](https://nexte.st/book), and while we don't need a lot of its features, it does have a couple of them that may benefit us. For instance, it allows retries with backoff, test timeouts, and test reporting (using junit format). 

## Changes

- adds `nextest` as the test runner on GH Actions (for Rust only, obviously).
- enable retries with exponential backoff
- automate creating issues for flaky tests.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### CODE REVIEW

**Comment by agourlay** [comment]
It is interesting that the flaky test `segment_builder_test::test_building_cancellation` fails consistently for me locally with `nextest`.

However it fails only with `cargo nextest run --all` 

```
thread 'segment_builder_test::test_building_cancellation' panicked at 'time_fast: 987, time_long: 501, is_stopped_long: true', lib/segment/tests/integration/segment_builder_test.rs:146:5
```

but NOT in isolation with `cargo nextest run --all test_building_cancellation`

```
cargo nextest run --all test_building_cancellation
    Finished test [unoptimized + debuginfo] target(s) in 0.24s
    Starting 1 tests across 8 binaries (330 skipped)
        PASS [   1.348s] segment::integration segment_builder_test::test_building_cancellation
------------
     Summary [   1.349s] 1 tests run: 1 passed, 330 skipped
```

File: .github/workflows/rust-lint.yml [hunk 1]
```diff
@@ -18,7 +18,6 @@ jobs:
     - name: Install miniml nightly (only for fmt)
       uses: dtolnay/rust-toolchain@nightly
       with:
-        profile: minimal
```
Comment (coszio) [inline .github/workflows/rust-lint.yml:21]: I deleted this because it is not an available option anymore

**Comment by coszio** [comment]
Tests aren't passing because (as @agourlay said) `test_building_cancellation` fails much more consistently when using `nextest`. This has been already addressed in #2094, but additional insight is needed to review it.

That said, github issue automation has been already tested on a personal repo, and [creates issues like this](https://github.com/coszio/gh-actions-playground/issues/3):

imagen

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -26,12 +25,63 @@ jobs:
...
+    - name: Install nextest
+      uses: taiki-e/install-action@nextest
+    - name: Build and run tests
+      # Profile "ci" is configured in .config/nextest.toml
+      run: cargo nextest run --workspace --profile ci
+    - name: Upload test report
+      uses: actions/upload-artifact@v2
+      with:
+        name: junit-${{ matrix.os }}.xml
+        path: target/nextest/ci/junit.xml
```
Comment (timvisee) [inline .github/workflows/rust.yml:37]: Is this is a proper and full replacement for `cargo test --all`? Or do we better create another job just for nextest runs?

Comment (coszio) [reply .github/workflows/rust.yml:37]: Yes, actually, the `--all` flag has been deprecated in favor of `--workspace`. It is full replacement 👍

**Comment by agourlay** [comment]
@coszio I am interested in seeing this land.

Please rebase :)

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -26,12 +25,63 @@ jobs:
...
+        if: ${{ steps.process-test-report.outputs.has_flaky_tests == 'true' }}
+        run: |
+          echo "Flaky tests found"
+          echo test=$(jq '.["@name"]' flaky_tests.json -r ) >> $GITHUB_OUTPUT
+          delimiter="###r###" 
+          echo "content<<$delimiter" >> $GITHUB_OUTPUT
+          echo "$(jq '[.flakyFailure] | flatten | .[0]["system-err"]' flaky_tests.json -r)" >> $GITHUB_OUTPUT
+          echo $delimiter >> $GITHUB_OUTPUT
+      - uses: actions/checkout@v3
+        if: ${{ steps.process-test-report.outputs.has_flaky_tests == 'true' }}
```
Comment (ffuugoo) [inline .github/workflows/rust.yml:70]: Why do we need a checkout here? 🤔

Comment (coszio) [reply .github/workflows/rust.yml:70]: We only need it for fetching the `flaky_test.md` issue template

**Comment by coszio** [comment]
Ok, now it has been rebased and is ready for another review pass

**Comment by agourlay** [comment]
With `nextest` locally I always get

```
--- STDERR:              collection common::stoppable_task::tests::test_task_panic ---
thread 'tokio-runtime-worker' panicked at 'stoppable task panicked', lib/collection/src/common/stoppable_task.rs:206:17
stack backtrace:
thread 'common::stoppable_task::tests::test_task_panic' panicked at 'assertion failed: handle.is_finished()', lib/collection/src/common/stoppable_task.rs:219:9
   0:     0x55e9820ce681 - std::backtrace_rs::backtrace::libunwind::trace::he648b5c8dd376705
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x55e9820ce681 - std::backtrace_rs::backtrace::trace_unsynchronized::h5da3e203eef39e9f
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x55e9820ce681 - std::sys_common::backtrace::_print_fmt::h8d28d3f20588ae4c
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x55e9820ce681 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hd9a5b0c9c6b058c0
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x55e9820ffaaf - core::fmt::rt::Argument::fmt::h0afc04119f252b53
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/fmt/rt.rs:138:9
   5:     0x55e9820ffaaf - core::fmt::write::h50b1b3e73851a6fe
```

Anyone else?

**Comment by timvisee** [comment]
> With `nextest` locally I always get
> 
> ```
> --- STDERR:              collection common::stoppable_task::tests::test_task_panic ---
> thread 'tokio-runtime-worker' panicked at 'stoppable task panicked', lib/collection/src/common/stoppable_task.rs:206:17
> stack backtrace:
> thread 'common::stoppable_task::tests::test_task_panic' panicked at 'assertion failed: handle.is_finished()', lib/collection/src/common/stoppable_task.rs:219:9
>    0:     0x55e9820ce681 - std::backtrace_rs::backtrace::libunwind::trace::he648b5c8dd376705
>                                at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
>    1:     0x55e9820ce681 - std::backtrace_rs::backtrace::trace_unsynchronized::h5da3e203eef39e9f
>                                at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
>    2:     0x55e9820ce681 - std::sys_common::backtrace::_print_fmt::h8d28d3f20588ae4c
>                                at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/sys_common/backtrace.rs:65:5
>    3:     0x55e9820ce681 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hd9a5b0c9c6b058c0
>                                at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/sys_common/backtrace.rs:44:22
>    4:     0x55e9820ffaaf - core::fmt::rt::Argument::fmt::h0afc04119f252b53
>                                at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/fmt/rt.rs:138:9
>    5:     0x55e9820ffaaf - core::fmt::write::h50b1b3e73851a6fe
> ```
> 
> Anyone else?

For me `cargo nextest run --all` works with no errors.

**Comment by agourlay** [comment]
@coszio Can you please rebase on dev, it seems to fail only on this branch for me.

**Comment by agourlay** [comment]
Still failing for me, I think those time based tests are just to brittle for system which are saturating.

I am afraid that one day they start failing on CI as well.

**Review by agourlay — APPROVED** [review]
I am fine to give this a try :+1: 

We can always revert in a few weeks if it is causes too many problems down the line.

**Comment by coszio** [comment]
> I am afraid that one day they start failing on CI as well.

My take here is that if tests are brittle, we should find a way to improve them. However, by using this changes, we won't have the pain of rerunning jobs when tests are flaky, and we'll still keep track of them via Github Issues

File: .github/workflows/rust.yml [hunk 4]
```diff
@@ -26,12 +25,71 @@ jobs:
...
+    - name: Upload test report
+      uses: actions/upload-artifact@v3
+      with:
+        name: junit-${{ matrix.os }}.xml
+        path: target/nextest/ci/junit.xml
+        
+  # After tests are run, this hacky script will process the JUnit output of nextest
+  # and will create a GH Issue if there is a test marked as flaky
+  process-results:
+    runs-on: ubuntu-latest
```
Comment (timvisee) [inline .github/workflows/rust.yml:42]: Could we incorporate this in the test job? That would prevent spawning 3 extra jobs just to process test failures.

Comment (coszio) [reply .github/workflows/rust.yml:42]: Unfortunately the script doesn't work well in windows

Comment (timvisee) [reply .github/workflows/rust.yml:42]: I see.

In that case, lets experiment with this set up and see how we like it. Maybe we can improve upon this in the future if it becomes a problem.

**Review by timvisee — APPROVED** [review]

**Pull request merged**